src/Symfony/Component/Security/Ldap/Ldap.php

```diff
@@ -0,0 +1,370 @@
+     */
+    public function getUseSsl()
+    {
+        return $this->useSsl;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setUseSsl($useSsl)
+    {
+        $this->useSsl = (boolean) $useSsl;
+
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getUseStartTls()
+    {
+        return $this->useStartTls;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setUseStartTls($useStartTls)
+    {
+        $this->useStartTls = (boolean) $useStartTls;
+
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getOptReferrals()
+    {
+        return $this->optReferrals;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setOptReferrals($optReferrals)
+    {
+        $this->optReferrals = (boolean) $optReferrals;
+
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getUsername()
+    {
+        return $this->username;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setUsername($username)
+    {
+        $this->username = $username;
+
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getPassword()
+    {
+        return $this->password;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setPassword($password)
+    {
+        $this->password = $password;
+
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getUsernameWithSuffix($username = null)
+    {
+        if (null === $username) {
+            $username = $this->username;
+        }
+
+        return $username.$this->usernameSuffix;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function findByUsername($username, $query, $filter = '*')
+    {
+        if (!$this->connection) {
+            $this->connect();
+        }
+
+        if (!is_array($filter)) {
+            $filter = array($filter);
+        }
+
+        $username = $this->escapeValue($this->getUsernameWithSuffix());
+
+        $query  = sprintf($query, $username);
+        $search = ldap_search($this->connection, $this->getDn(), $query, $filter);
+        $infos  = ldap_get_entries($this->connection, $search);
+
+        if (0 === $infos['count']) {
+            return null;
+        }
+
+        return $infos[0];
+    }
+
+    private function connect()
+    {
+        if (!$this->connection) {
+            $host = $this->getHost();
+            if ($this->getUseSsl()) {
+                $host = 'ldaps://' . $host;
+            }
+            ldap_set_option($this->connection, LDAP_OPT_PROTOCOL_VERSION, $this->getVersion());
+            ldap_set_option($this->connection, LDAP_OPT_REFERRALS, $this->getOptReferrals());
+
+            if ($this->getUseStartTls()) {
+                ldap_start_tls($this->connection);
+            }
+
+            $this->connection = ldap_connect($host, $this->getPort());
```

Comment:
You should do the `ldap_connect()` first, then do the `ldap_set_option()` and `ldap_start_tls()`.